diff -Naur hybrid-v35_64-nodebug-pcoem-6_30_223_271.orig/src/include/linuxver.h hybrid-v35_64-nodebug-pcoem-6_30_223_271/src/include/linuxver.h
--- hybrid-v35_64-nodebug-pcoem-6_30_223_271.orig/src/include/linuxver.h	2015-09-19 00:47:30.000000000 +0200
+++ hybrid-v35_64-nodebug-pcoem-6_30_223_271/src/include/linuxver.h	2017-11-24 22:14:31.528878919 +0100
@@ -438,7 +438,7 @@
 #define SET_NETDEV_DEV(net, pdev)	do {} while (0)
 #endif
 
-#ifndef HAVE_FREE_NETDEV
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(3, 1, 0) && !defined(HAVE_FREE_NETDEV))
 #define free_netdev(dev)		kfree(dev)
 #endif
 
